api/metrics/v1/rules.go

```diff
@@ -107,7 +114,56 @@ func (rh *rulesHandler) put(w http.ResponseWriter, r *http.Request) {
 		http.Error(w, "error finding tenant", http.StatusUnauthorized)
 	}
 
-	resp, err := rh.client.SetRulesWithBody(r.Context(), tenant, r.Header.Get("Content-type"), r.Body)
+	id, ok := authentication.GetTenantID(r.Context())
+	if !ok {
+		http.Error(w, "error finding tenant ID", http.StatusUnauthorized)
+		return
+	}
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, "error reading request body", http.StatusInternalServerError)
+		return
+	}
+
+	var rawRules rules.Rules
+	if err := yaml.Unmarshal(body, &rawRules); err != nil {
+		level.Error(rh.logger).Log("msg", "could not unmarshal rules", "err", err.Error())
+		http.Error(w, "error unmarshaling rules", http.StatusInternalServerError)
+
+		return
+	}
+
+	for i := range rawRules.Groups {
+		for j := range rawRules.Groups[i].Rules {
+			switch r := rawRules.Groups[i].Rules[j].(type) {
+			case rules.RecordingRule:
+				if r.Labels.AdditionalProperties == nil {
+					r.Labels.AdditionalProperties = make(map[string]string)
+				}
+
+				r.Labels.AdditionalProperties[rh.tenantLabel] = id
+				rawRules.Groups[i].Rules[j] = r
+			case rules.AlertingRule:
+				if r.Labels.AdditionalProperties == nil {
+					r.Labels.AdditionalProperties = make(map[string]string)
+				}
+
+				r.Labels.AdditionalProperties[rh.tenantLabel] = id
+				rawRules.Groups[i].Rules[j] = r
+			}
+		}
+	}
+
+	body, err = yaml.Marshal(rawRules)
+	if err != nil {
+		level.Error(rh.logger).Log("msg", "could not marshal YAML", "err", err.Error())
+		http.Error(w, "error marshaling YAML", http.StatusInternalServerError)
```

Comment:
```suggestion
		http.Error(w, "error marshaling rules YAML", http.StatusInternalServerError)
```

Reply:
I think since we have the same marshaling logic also in the `get` [method](https://github.com/observatorium/api/blob/80ff927bebedc78bf7092d494526bd88a5428b3b/api/metrics/v1/rules.go#L97-L103) maybe we could extract this to a separate function and reuse it in both get/put endpoints? same for the [auth logic](https://github.com/observatorium/api/blob/80ff927bebedc78bf7092d494526bd88a5428b3b/api/metrics/v1/rules.go#L23-L33)

Reply:
Yes, I think we can do it for both marshaling and unmarshalling rules!